### ISSUE TITLE
Plans for libhsdaoh: SDR or multi-purpose?

### ISSUE THREAD

**Issue by Stefan-Olt** [steve-m/hsdaoh#1]
Hi,
as I would like to use hsdaoh for MISRC the questions is: Is libhsdaoh going to be a general-purpose library designed for handling the data produced by the hsdaoh FPGA code and interfacing the HDMI capture devices or is it specifically targeting SDR applications in the future, because the SDR branch contains SDR-specific functions, like data conversion to float and DC balance, but also setting sample rates (that's also something good for MISRC, but I have to use the external PLL as the internal one cannot generate 40 MHz with the 27 MHz reference).
In this case I would just copy libhsdaoh and modify it for my needs. But if the SDR branch is just an experiment and the goal is for example to have a general-purpose libhsdaoh, while the SDR-specific processing is done in a libhsdaohsdr, I would try to use an unmodified libhsdaoh and create pull requests if needed (for example handing over the `pack_state` to the callback function).

I would prefer the second approach, as this would make the entire idea more universal (other use case could be a logic analyzer for sigrok) and adding for example new HDMI capture devices in the future could be done in a single place and would not require to back-port any changes.

Best regards
Stefan

**Comment by steve-m** [comment]
The idea was for it to be a general purpose library, and the sdr branch is more or the less a hack. So yeah, the second approach would be the cleanest.

Btw, which FPGA board are you using? Since you're talking about the external PLL I assume the nano 20K.
With my tests the external PLL had more spurs than the internal PLL unfortunately.
Would be nice if the CLKDIV allowed /3, then you could use the 120 MHz.

**Comment by Stefan-Olt** [comment]
That's great to here. I'm planning to use the 20K, because I need 160 MB/s and it just has enough pins for 32 Bit input (there are some additional on the sd card that are not on the pin headers). It has not yet arrived.

Maybe I'll use a fixed external clock then if the PLL is not that great

**Comment by Stefan-Olt** [comment]
Do you see any chance that the ideas of [my fork](https://github.com/Stefan-Olt/hsdaoh)  can be integrated:
- Allow a "raw callback", so that the application needs to handle processing the frame: As misrc_capture is multithreaded and using a ringbuffer, it is more efficient to copy the frame data directly into it rather than to have a another buffer in between. Some internal functions need to be exposed to the API for that (metadata extraction, CRC check)
- Allow a custom message handler. I have redesigned my original approach into a fprint-like approach, making it easier to use.

The "raw callback" can have another benefit: As you provide a modified firmware for the MS2130, it should be possible to use standard OS api to open the device for capture, that would make it easier to use on Windows (no Zadig anymore, MS2130 can also be used for other capture) and no sudo required on Mac (since macOS 12 you cannot use libusb anymore for a device with standard drivers without root. A signed GUI application from Apple registered developers can bypass that by registering that right in the OS somehow, but I'm not going to do that). So the data processing is separate from the device handling.
Additionally there might be the option to use a MIPI as alternative HDMI and the Rockchip RK3588 has a HDMI input, there are several small boards that use this chip and have a HDMI input. Maybe this could also be used, not sure if it modifies any data. I had a quick look at the kernel driver for it, I couldn't find anything obvious related to that.

**Comment by steve-m** [comment]
So let me summarize: you both want to use a raw callback and don't want use the input via libuvc but direct OS APIs

Then what's the point of using this library altogether? There isn't much code left that you are using.
You just could copy the metadata header and implement everything in misrc_capture and be done with it...
I'm sure you'll also find a way to do things like the CRC more efficiently with some x86 assembly anyways.

**Comment by Stefan-Olt** [comment]
No, I want different options in misrc_capture, ms2130+libuvc being one of them, using the OS APIs another (I wouldn't use the different OS apis but rather ffmpeg/libavdevice because I already link to ffmpeg anyway).
Therefore I would like to be able to use hsdaoh in a raw mode to use the same callback.

Not sure if CRC is really the bottleneck, maybe on a Raspberry Pi. I found [this comparing 3 different types of CRC16 CCITT implementations](https://github.com/mattsta/crcspeed), the fastest method being 4 times faster than a lookup table, all plain C.

Not sure if you had a look at my fork, it also contains fixes to be able to build hsdaoh again on Mac (macOS has different endian conversion functions than Linux)
